enhancements/deprecated-bundles.md

```diff
@@ -0,0 +1,117 @@
+- Alerts are fired if terminated bundles exist on cluster
+- OLM prevents cluster upgrades based on whether versions are terminated in the upgraded cluster’s index
+
+## Proposal
+
+We propose to add a command to `opm index` called `deprecate` which takes in a set of bundle images, an index, and all the same arguments as `opm index add`. The command modifies the index by removing all versions that come after the specified bundle image in the update graph. In this way, versions that were removed from the index will not receive updates if subscriptions still exist for them on cluster. A new field is added to the operator-registry database table `operatorbundle` called `deprecated` and the bundle images specified in the command will be marked as such.
+
+The operator-registry APIs will be modified to ensure that OLM does not allow install of deprecated bundles. This also means appropriate status reporting in the subscription and installplan need to be added to OLM, and alerts will be fired if a deprecated version is installed.
+
+### User Stories
+
+#### Story 1
+
+As an index maintainer I want to be able to truncate update graphs in order to remove or prevent installs of incompatible or unsupported versions of certain operators.
+
+##### Implementation Details
+
+A new field is added to the operator-registry database table `operatorbundle` called `deprecated` and the bundle images specified in the command will be marked as such. Versions that come after the deprecated version in the update graph (in all channels) are removed from the database.
+
+For example, suppose following channel:
+
+```
+    1.4.0 -- replaces -> 1.3.0 -- replaces -> 1.2.0 -- replaces -> 1.1.0
+```
+
+we then deprecate version 1.3.0 using the following command
+
+```
+    $ opm index deprecate --bundles "quay.io/my/bundle:1.3.0" --from-index "quay.io/my/index:v4.6" --tag "quay.io/my/index:v4.7"
+```
+
+Now, in `quay.io/my/index:v4.7` the above channel looks like:
+
+```
+    1.4.0 -- replaces -> 1.3.0 [deprecated]
+```
+
+Deprecated versions are upgradable but not installable. To achieve this, we will modify the operator-registry apis used by OLM to determine what can be installed.
+
+#### Story 2
+
+As an index maintainer I want to be able to undo the above operation ideally without removing the entire package and re-adding all the bundles.
+
+##### Implementation Details
+
+The `opm index rm` command is modified to allow removal of the tail a channel (deprecated versions will always lie at the tail of the graph).
+
+The `opm index add` command is modified to allow adding to the tail of a channel (removed versions from the deprecate command will always have been from the tail)
+
+With these two additions we can now undo the above deprecation by:
+
+- removing 1.3.0 (this is mainly to remove the deprecation notice)
+- adding 1.3.0, 1.2.0, 1.1.0
+
+#### Story 3
+
+As a cluster admin, I want to be alerted if subscriptions exist on cluster for deprecated versions of operators.
```

Comment:
Isn't this called out as a non-goal

Reply:
the non-goal is for terminated bundles

Reply:
I think the alert should be for:

1. "Subscription is to a channel that no longer exists, or the head is deprecated"
2. "A Manual subscription is on a deprecated version"
3. "Stranded version is installed" we need to document how to get out of this situation by installing the older catalog

Also with the new work on the skip range being steppable, I imagine that this will be less likely to happen.